contrib/src/templates/mod.rs

```diff
@@ -215,13 +249,18 @@ impl Template {
                 }
             };
 
-            match Context::initialize(template_root) {
-                Some(mut ctxt) => {
-                    f(&mut ctxt.engines);
-                    Ok(rocket.manage(ctxt))
-                }
-                None => Err(rocket)
-            }
+            let ctxt = match Context::initialize(template_root, Box::new(f)) {
+                Some(ctxt) => ctxt,
+                None => return Err(rocket),
+            };
+
+            #[cfg(debug_assertions)]
+            let rocket = rocket.attach(AdHoc::on_request(|req, _| {
```

Comment:
I think we're better off creating a custom `Fairing` implementation at this point as opposed to using `AdHoc`. This is really confusing. It took me a while to realize that this is what's actually going to reload the templates.

I think we want to create a new file, `fairing.rs` or something, that contains `ManagedContext` and some new type, `TemplateFairing`, for which we implement `Fairing` for. This time it will have both `on_attach` and `on_request` handlers. The `custom` method should then get really really small, probably just: `TemplateFairing::new()`.